sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -735,33 +768,94 @@ static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
 	}
 }
 
+/*
+ * check endpoint status in slaves and gather link ID for all slaves in
+ * the same group to generate different CPU DAI. Now only support
+ * one sdw link with all slaves set with only single group id.
+ */
+static void get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
+			   int *dai_id, int *dai_num, int *codec_num,
+			   bool *group_generated)
+{
+	const struct snd_soc_acpi_adr_device *adr_d;
+	const struct snd_soc_acpi_link_adr *adr_next;
+	int group_id;
+	int index = 0;
+
+	*codec_num = adr_link->num_adr;
+	adr_d = adr_link->adr_d;
+
+	dai_id[index++] = ffs(adr_link->mask) - 1;
+	if (!adr_d->endpoints->aggregated) {
+		*dai_num = index;
+		return;
+	}
+
+	group_id = adr_d->endpoints->group_id;
+
+	/* gather other link ID of slaves in the same group */
+	for (adr_next = adr_link + 1; adr_next->num_adr;  adr_next++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
+
+		endpoints = adr_next->adr_d->endpoints;
+		if (!endpoints->aggregated ||
+		    endpoints->group_id != group_id)
+			continue;
+
+		dai_id[index++] = ffs(adr_next->mask) - 1;
+		*codec_num += adr_next->num_adr;
+	}
+
+	/*
+	 * indicate CPU DAIs for this group have been generated
+	 * to avoid generating CPU DAIs for this group again.
+	 */
+	group_generated[group_id - 1] = TRUE;
+
+	*dai_num = index;
+}
+
 static int create_sdw_codec_dai(struct device *dev,
 				struct snd_soc_dai_link *dai_links, int *id,
 				struct snd_soc_dai_link_component *cpus,
 				const struct snd_soc_acpi_link_adr *link,
-				int *cpus_id)
+				int *cpus_id, bool *group_generated)
 {
+	const struct snd_soc_acpi_link_adr *link_next;
 	struct snd_soc_dai_link_component *codec;
-	unsigned int part_id, link_id;
-	int i, j = 0, idx;
+	int cpu_dai_id[SDW_MAX_DEVICES];
+	int cpu_dai_num, cpu_dai_index;
+	unsigned int part_id, link_num;
+	int i, j = 0, k, idx;
 	int ret;
 
-	codec = devm_kcalloc(dev, link->num_adr,
-			     sizeof(struct snd_soc_dai_link_component),
-			     GFP_KERNEL);
+	get_slave_info(link, cpu_dai_id, &cpu_dai_num, &i, group_generated);
+	codec = devm_kcalloc(dev, i, sizeof(*codec), GFP_KERNEL);
 	if (!codec)
 		return -ENOMEM;
 
-	ret = create_codec_dai_name(dev, link, codec);
-	if (ret < 0)
-		return ret;
+	link_next = link;
+	link_num = cpu_dai_num;
+
+	/* generate codec name on different links in the same group */
+	for (i = 0, idx = 0; link_next->num_adr && i < link_num; link_next++) {
+		if (cpu_dai_id[i] != ffs(link_next->mask) - 1)
+			continue;
+
+		ret = create_codec_dai_name(dev, link_next, codec, idx);
+		if (ret < 0)
+			return ret;
+
+		i++;
+		idx += link_next->num_adr;
+	}
```

Comment:
@RanderWang this is really good work but it's very difficult to review.
You have too many indices and variables (id, idx, index, link, etc), that really makes it difficult to figure out how your solution works.

please see attached suggestions to simplify the code.
[sdw_machine.diff.txt](https://github.com/thesofproject/linux/files/4257833/sdw_machine.diff.txt)

Reply:
thanks very much! I will check the file and refine the PR.